service/matching/matcher.go

```diff
@@ -129,6 +154,19 @@ func (tm *TaskMatcher) Offer(ctx context.Context, task *internalTask) (bool, err
 	}
 }
 
+func (tm *TaskMatcher) offerOrTimeout(ctx context.Context, task *internalTask) (bool, error) {
+	select {
+	case tm.taskC <- task: // poller picked up the task
+		if task.responseC != nil {
+			err := <-task.responseC
```

Comment:
Do you need to honor the ctx deadline while waiting on response?

Reply:
Good question. From a behavior standpoint, if it takes too long for the response, the caller sees timeout anyway (irrespective of whether we return from here or not). This can happen when RecordDecisionTaskStarted() simply times out and is in a retry loop.  The only reason to do this would be for system protection i.e. when history is down and too may goroutines wait here for the response (for the duration of the retry/backoff expiration).  Since it doesn't change behavior for the caller (one who calls Offer()), I think it's worth honoring the deadline.